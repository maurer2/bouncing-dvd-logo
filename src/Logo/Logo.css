.logo {
  position: absolute;
  width: 10%;
  height: auto;
  top: 0;
  left: 0;
  transform: translateX(0%) translateY(0%);
  background: red;
}
