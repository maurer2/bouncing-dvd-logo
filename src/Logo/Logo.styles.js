import styled from 'styled-components/macro';

// eslint-disable-next-line import/prefer-default-export
export const LogoElement = styled.div.attrs((props) => ({
  style: {
    transform: `translate(${Math.round(props.positionX)}px, ${Math.round(props.positionY)}px)`,
  },
}))`
  position: absolute;
  top: 0;
  left: 0;
  width: ${(props) => `${props.widthValue}px`};
  height: ${(props) => `${props.heightValue}px`};
  color: ${(props) => `${props.colourValue}`};
  will-change: transform;
`;
