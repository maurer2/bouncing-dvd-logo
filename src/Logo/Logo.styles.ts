import styled from 'styled-components';

import type { LogoStyleProps } from './Logo.types';

export const LogoElement = styled.figure.attrs<LogoStyleProps>((props) => ({
  style: {
    transform: `translate(${Math.floor(props.$positionX)}px, ${Math.floor(props.$positionY)}px)`,
    color: props.$colour,
  },
}))<LogoStyleProps>`
  position: absolute;
  top: 0;
  left: 0;
  width: ${(props) => `${props.$width}px`};
  height: ${(props) => `${props.$height}px`};
  margin: 0;
  will-change: transform;
  opacity: ${({ $isPaused }) => ($isPaused ? '0.25' : '1')};
`;
