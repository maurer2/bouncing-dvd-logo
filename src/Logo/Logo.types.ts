export type LogoProps = {
  positionX: number;
  positionY: number;
  width: number;
  height: number;
  changeColours: boolean;
  isPaused: boolean;
}
