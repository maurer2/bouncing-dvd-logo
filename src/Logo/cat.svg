<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 727.27 672.28">
  <g transform="translate(-30.32 -318.67)">
    <g transform="matrix(6.8233 0 0 6.8233 -47.527 191.7)">
      <path style="stroke-linejoin:round; stroke:#ffffff; stroke-linecap:round; stroke-width:6; fill:none;" d="m15.454 21.608 4.879 25.876c-3.7786 6.3618-5.9245 13.63-5.9245 21.346 0 25.017 22.513 45.305 50.271 45.305s50.315-20.288 50.315-45.305c0-7.4745-2.0588-14.519-5.6196-20.736l4.879-25.833-20.562 9.5402c-8.1959-5.2251-18.217-8.3205-29.013-8.3205-10.481 0-20.213 2.8909-28.272 7.8413l-20.948-9.714z"/>
      <path style="stroke-linejoin:round; stroke:#000000; stroke-linecap:round; stroke-width:3.9782; fill:currentColor" d="m15.454 21.608 4.879 25.876c-3.7786 6.3618-5.9245 13.63-5.9245 21.346 0 25.017 22.513 45.305 50.271 45.305s50.315-20.288 50.315-45.305c0-7.4745-2.0588-14.519-5.6196-20.736l4.879-25.833-20.562 9.5402c-8.1959-5.2251-18.217-8.3205-29.013-8.3205-10.481 0-20.213 2.8909-28.272 7.8413l-20.948-9.714z"/>
      <path style="stroke-linejoin:round; stroke:#000000; stroke-linecap:round; stroke-width:2.788; fill:none" d="m41.227 86.168c0.71997 5.451 3.9594 11.035 11.152 11.152 9.5596 0.15794 17.384-10.258 19.682-17.547-5.2586-0.76083-10.517-0.71929-15.776-0.23408 2.2988 6.9385 10.416 17.652 18.397 17.781 7.6568 0.12382 10.288-4.9103 11.152-11.152"/>
      <path style="fill:#000000" d="m76 48c0 2.2091-1.7909 4-4 4s-4-1.7909-4-4 1.7909-4 4-4 4 1.7909 4 4z" transform="matrix(1.9388 .87230 -.87230 1.9388 -50.299 -95.029)"/>
      <path style="fill:#ffffff" d="m76 48c0 2.2091-1.7909 4-4 4s-4-1.7909-4-4 1.7909-4 4-4 4 1.7909 4 4z" transform="matrix(.76796 .34552 -.34552 .76796 11.205 -1.3029)"/>
      <path style="fill:#000000" d="m76 48c0 2.2091-1.7909 4-4 4s-4-1.7909-4-4 1.7909-4 4-4 4 1.7909 4 4z" transform="matrix(1.9388 .87230 -.87230 1.9388 -15.389 -95.1)"/>
      <path style="fill:#ffffff" d="m76 48c0 2.2091-1.7909 4-4 4s-4-1.7909-4-4 1.7909-4 4-4 4 1.7909 4 4z" transform="matrix(.76796 .34552 -.34552 .76796 46.115 -1.3742)"/>
    </g>
  </g>
</svg>
