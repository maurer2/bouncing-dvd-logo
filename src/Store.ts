import { createContext } from 'react';

const colours = ['white', 'red', 'blue', 'yellow', 'fuchsia', 'lime'] as const;
type Colours = typeof colours[number];

export type StoreType = {
  colours: Colours[];
  soundIsDisabled: boolean;
  toggleSound: () => void;
};

const Store = createContext<StoreType | null>(null);

export default Store;
export {
  colours,
};
