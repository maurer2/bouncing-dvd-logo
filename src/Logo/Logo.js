import React, { Component } from 'react';
import './Logo.css';
import { ReactComponent as CatSVG } from './cat.svg';

class Logo extends Component {
  constructor(props) {
    super(props);

    this.getStyle = this.getStyle.bind(this);
  }

  getStyle() {

    return {
      // transform: `translate(calc(-50% + ${this.props.positionX}vw), calc(-50% + ${this.props.positionY}vw))`,
      transform: `translate(${this.props.positionX}px, ${this.props.positionY}px)`,
      width: `${this.props.width}px`
    };
  }

  render() {
    return (
      <CatSVG className="logo" style= { this.getStyle() } />
    );
  }
}

export default Logo;
