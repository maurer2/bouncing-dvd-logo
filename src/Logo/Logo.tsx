import React, {
  useEffect, useRef, useContext, VFC,
} from 'react';
import PropTypes from 'prop-types';

import Store from '../Store';
import useColour from '../Hooks/useColour';
import { ReactComponent as CatLogo } from '../assets/cat.svg';

import * as Styles from './Logo.styles';
import * as Types from './Logo.types';

const Logo: VFC<Readonly<Types.LogoProps>> = ({
  positionX,
  positionY,
  width,
  height,
  changeColours,
  isPaused,
}): JSX.Element => {
  const { colours } = useContext(Store);
  const prevChangeColours = useRef(false);
  const [colour, setColour] = useColour(colours);

  useEffect(() => {
    if (changeColours && changeColours !== prevChangeColours.current) {
      setColour();
    }

    prevChangeColours.current = changeColours;
  }, [changeColours, setColour]);

  return (
    <Styles.LogoElement
      $positionX={positionX}
      $positionY={positionY}
      $width={width}
      $height={height}
      $colour={colour}
      $isPaused={isPaused}
      data-testid="logo-element"
    >
      <CatLogo data-testid="cat-logo" />
    </Styles.LogoElement>
  );
};

const { number, bool } = PropTypes;

Logo.propTypes = {
  positionX: number.isRequired,
  positionY: number.isRequired,
  width: number.isRequired,
  height: number.isRequired,
  changeColours: bool.isRequired,
  isPaused: bool.isRequired,
};

export default Logo;
