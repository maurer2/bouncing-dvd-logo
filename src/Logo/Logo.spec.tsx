import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import Store from '../Store';

import Component from './Logo';
import { LogoProps } from './Logo.types';

describe('Logo', () => {
  const storeValues = { colours: ['white', 'red', 'blue', 'yellow', 'fuchsia', 'lime'] };

  const defaultProps: LogoProps = {
    positionX: 100,
    positionY: 100,
    width: 100,
    height: 100,
    changeColours: false,
    isPaused: false,
  };

  const triggerColourChangeProp: LogoProps = {
    ...defaultProps,
    changeColours: true,
  };

  const StoreProvider = ({ children }) => (
    <Store.Provider value={storeValues}>{children}</Store.Provider>
  );

  const setup = (props) => render(
    <StoreProvider key={props?.key}>
      <Component {...defaultProps} {...props} />
    </StoreProvider>,
  );

  it('should render ', () => {
    const screen = setup({});

    expect(screen.getByTestId('logo-element')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const screen = setup({});

    expect(screen.container.firstChild).toMatchSnapshot();
  });

  it('should have child elements', () => {
    const screen = setup({});

    expect(screen.getByTestId('logo-element')).toBeInTheDocument();
    expect(screen.getByTestId('cat-logo')).toBeTruthy();
  });

  it('should have initial colour', () => {
    const screen = setup({});

    expect(screen.getByTestId('logo-element').style.getPropertyValue('color')).toBe(
      storeValues.colours[0],
    );
    expect(screen.getByTestId('logo-element')).toHaveStyle(`color: ${storeValues.colours[0]}`);
  });

  it('should change colour', async () => {
    const { rerender, getByTestId } = setup({});
    const startColour = getByTestId('logo-element').style.getPropertyValue('color');

    rerender(
      <StoreProvider>
        <Component {...defaultProps} {...triggerColourChangeProp} />
      </StoreProvider>,
    );

    const coloursWithoutStartColour = storeValues.colours.filter(
      (colour) => colour !== startColour,
    );
    const currentColour = getByTestId('logo-element').style.getPropertyValue('color');

    expect(coloursWithoutStartColour).toContain(currentColour);
    expect(coloursWithoutStartColour).not.toContain(startColour);

    expect(getByTestId('logo-element').style.getPropertyValue('color')).not.toBe(
      storeValues.colours[0],
    );
    expect(getByTestId('logo-element')).toHaveStyle(`color: ${currentColour}`);
  });
});
