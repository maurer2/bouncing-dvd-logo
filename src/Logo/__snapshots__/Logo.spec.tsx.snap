// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logo should match snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100px;
  height: 100px;
  margin: 0;
  will-change: transform;
  opacity: 1;
}

<figure
  class="c0"
  data-testid="logo-element"
  style="transform: translate(100px, 100px); color: white;"
>
  <svg
    data-testid="cat-logo"
  >
    SVG markup
  </svg>
</figure>
`;
